python2-pybind11/challenge/basics.cpp

```diff
@@ -0,0 +1,39 @@
+#include <pybind11/pybind11.h>
+
+#include "basics.hpp"
+
+namespace py = pybind11;
+
+PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
+
+namespace basics {
+
+PYBIND11_PLUGIN(basics) {
+    py::module m("basics", "wrapped C++ basics module");
+
+    m.def("compare", &compare);
+    m.def("adjacent", &adjacent);
+
+    py::class_<Secret>(m, "Secret");
+
+    py::class_<Doodad, std::shared_ptr<Doodad>>(m, "Doodad")
+        .def(py::init<const std::string &, int>(), py::arg("name"), py::arg("value") = 1)
+        .def("__init__",
+            [](Doodad &instance, std::pair<std::string, int> p) {
+                new (&instance) Doodad(WhatsIt{p.first, p.second});
+            }
+        )
+        .def_readwrite("name", &Doodad::name)
+        .def_readwrite("value", &Doodad::value)
+        .def_static("get_const", &Doodad::get_const) // does not preserve const :(
+
+        // no mapping for unique_ptr but shared_ptr<Doodad> maps to Python Doodad directly
+        .def("clone", [](const Doodad &d) { return std::shared_ptr<Doodad>(d.clone().release()); })
```

Comment:
Thanks, fixed (with new changes landing [here](https://github.com/TallJimbo/python-cpp-challenge/tree/python2-pybind11).

I appreciate you paying attention to our little comparison experiment.  I'm personally quite impressed with pybind11 so far (but I'm far from the only person whose opinion matters).